core/templates/pages/exploration-player-page/services/diagnostic-test-player-engine.service.ts

```diff
@@ -0,0 +1,419 @@
+    private answerClassificationService: AnswerClassificationService,
+    private audioTranslationLanguageService: AudioTranslationLanguageService,
+    private contextService: ContextService,
+    private explorationHtmlFormatterService: ExplorationHtmlFormatterService,
+    private expressionInterpolationService: ExpressionInterpolationService,
+    private focusManagerService: FocusManagerService,
+    private questionBackendApiService: QuestionBackendApiService,
+    private diagnosticTestPlayerStatusService:
+      DiagnosticTestPlayerStatusService) {
+    this._numberOfAttemptedQuestions = 0;
+    this._encounteredQuestionIds = [];
+  }
+
+  init(
+      diagnosticTestTopicTrackerModel: DiagnosticTestTopicTrackerModel,
+      successCallback: (initialCard: StateCard, nextFocusLabel: string) => void
+  ): void {
+    this._diagnosticTestTopicTrackerModel = diagnosticTestTopicTrackerModel;
+    this._initialCopyOfTopicTrackerModel = cloneDeep(
+      diagnosticTestTopicTrackerModel);
+
+    this._currentTopicId = (
+      this._diagnosticTestTopicTrackerModel.selectNextTopicIdToTest());
+
+    this.questionBackendApiService.fetchDiagnosticTestQuestionsAsync(
+      this._currentTopicId, this._encounteredQuestionIds).then((response) => {
+      this._diagnosticTestCurrentTopicStatusModel = (
+        new DiagnosticTestCurrentTopicStatusModel(response));
+      this._currentSkillId = (
+        this._diagnosticTestCurrentTopicStatusModel.getNextSkill());
+      this._currentQuestion = (
+        this._diagnosticTestCurrentTopicStatusModel.getNextQuestion(
+          this._currentSkillId)
+      );
+      const stateCard = this.createCard(this._currentQuestion);
+      successCallback(stateCard, this._focusLabel);
+    });
+  }
+
+  private _getNextQuestion(
+      successCallback: (value: Question) => void,
+      errorCallback: () => void
+  ): void {
+    let currentTopicIsCompletelyTested = (
+      this._diagnosticTestCurrentTopicStatusModel.isTopicCompletelyTested());
+
+    if (currentTopicIsCompletelyTested) {
+      let topicIsPassed = (
+        this._diagnosticTestCurrentTopicStatusModel.isTopicPassed());
+
+      if (topicIsPassed) {
+        this._diagnosticTestTopicTrackerModel.recordTopicPassed(
+          this._currentTopicId);
+      } else {
+        this._diagnosticTestTopicTrackerModel.recordTopicFailed(
+          this._currentTopicId);
+      }
+
+      if (this.isDiagnosticTestFinished()) {
+        errorCallback();
+      }
+
+      this._currentTopicId = (
+        this._diagnosticTestTopicTrackerModel.selectNextTopicIdToTest());
+
+      this.questionBackendApiService.fetchDiagnosticTestQuestionsAsync(
+        this._currentTopicId, this._encounteredQuestionIds).then(
+        skillIdToQuestionsModel => {
+          this._diagnosticTestCurrentTopicStatusModel = (
+            new DiagnosticTestCurrentTopicStatusModel(
+              skillIdToQuestionsModel)
+          );
+
+          if (
```

Comment:
Why would the lifeline be consumed if we have only just loaded the new topic? (If this is actually a valid use case, maybe worth adding a comment to explain.)

Reply:
Removed